### PR TITLE
[Notifier] Add Firebase bridge

### PR DESCRIPTION
**Pull request by Jeroeny**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | See #33687
| License       | MIT

This would add [Firebase](https://firebase.google.com) integration for the Notifier component. With Firebase you can send push notifications to the users of you Android and iOS app and website (formerly known as Google Cloud messaging).

I'm not sure if it's possible to have this merged, like the other bridges. Or if I should create a stand-alone repository? That'd be fine too.

Also it's now using the `ChatMessage` as implementation of `Symfony\Component\Notifier\Message\MessageInterface`, but I feel like this component could use a `PushMessage` or something similar. Although I'm not sure if it would contain more than `subject` that the `ChatMessage` does.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return sprintf('firebase://%s', $this->getEndpoint());
+    }
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof ChatMessage;
+    }
+
+    /**
+     * @see https://core.telegram.org/bots/api
```
Comment (noniagriconomie) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php]: typo?

File: src/Symfony/Component/Notifier/Bridge/Firebase/Notification/IOSNotification.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Bridge\Firebase\Notification;
+
+use Symfony\Component\Notifier\Bridge\Firebase\FirebaseOptions;
+
+final class IOSNotification extends FirebaseOptions
+{
+    public function androidChannelId(string $androidChannelId): self
```
Comment (noniagriconomie) [inline src/Symfony/Component/Notifier/Bridge/Firebase/Notification/IOSNotification.php]: IOS but the method is android? is it a typo?

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php (hunk 1 above)
Comment (Jeroeny) [reply src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php]: Oops, thanks. Fixed both.

File: src/Symfony/Component/Notifier/Bridge/Firebase/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.0.0
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/CHANGELOG.md]: 5.1

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseOptions.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @internal
+ *
+ * @see https://firebase.google.com/docs/cloud-messaging/xmpp-server-ref.html
+ *
+ * @experimental in 5.0
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseOptions.php]: 5.1

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseOptions.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+namespace Symfony\Component\Notifier\Bridge\Firebase;
+
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @internal
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseOptions.php]: AFAIK this is not correct

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+use Symfony\Component\Notifier\Message\ChatMessage;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @internal
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php]: AFAIK this is not correct

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @experimental in 5.0
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php]: 5.1

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransportFactory.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+
+/**
+ * @author Jeroen Spee <spee.jeroen@gmail.com>
+ *
+ * @experimental in 5.0
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransportFactory.php]: 5.1

File: src/Symfony/Component/Notifier/Bridge/Firebase/Notification/AndroidNotification.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Bridge\Firebase\Notification;
+
+use Symfony\Component\Notifier\Bridge\Firebase\FirebaseOptions;
+
+final class AndroidNotification extends FirebaseOptions
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/Notification/AndroidNotification.php]: Please add experimental

File: src/Symfony/Component/Notifier/Bridge/Firebase/Notification/AndroidNotification.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+{
+    public function androidChannelId(string $androidChannelId): self
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/Notification/AndroidNotification.php]: channelId would be enough as method and variable name

File: src/Symfony/Component/Notifier/Bridge/Firebase/Notification/IOSNotification.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Bridge\Firebase\Notification;
+
+use Symfony\Component\Notifier\Bridge\Firebase\FirebaseOptions;
+
+final class IOSNotification extends FirebaseOptions
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/Notification/IOSNotification.php]: Please add experimental

File: src/Symfony/Component/Notifier/Bridge/Firebase/Notification/WebNotification.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Bridge\Firebase\Notification;
+
+use Symfony\Component\Notifier\Bridge\Firebase\FirebaseOptions;
+
+final class WebNotification extends FirebaseOptions
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/Notification/WebNotification.php]: Please add experimental

File: src/Symfony/Component/Notifier/Bridge/Firebase/composer.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+            "name": "Jeroen Spee",
+            "email": "spee.jeroen@gmail.com"
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": "^7.2.9",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Firebase/composer.json]: 7.2.5

File: src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransportFactory.php [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        $scheme = $dsn->getScheme();
+        $token = sprintf('%s:%s', $this->getUser($dsn), $this->getPassword($dsn));
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $port = $dsn->getPort();
+
+        if ('firebase' === $scheme) {
+            return (new FirebaseTransport($token, $this->client, $this->dispatcher))
+                ->setHost($host)
+                ->setPort($port);
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransportFactory.php]: should be on one line.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Jeroeny.

**Comment by acasademont** [comment]
After seeing fabpot's talk today talking about the Firebase notifier coming in 5.1 I looked at the repo and noticed that this Firebase notifier is based on the old Http API, that allows quite a lot less options than the new one, specifically, the new one allows you to send in 1 message different options for different platforms (iOS, Android, Web) instead of having to send separate messages for each of the platforms.

Not sure @Jeroeny if this is something that should be changed before the 5.1 release?

https://firebase.google.com/docs/cloud-messaging/migrate-v1

**Comment by fabpot** [comment]
@acasademont I would recommend to open a new issue.

**Pull request merged**